File: C++/0539.cpp
#include <algorithm>
#include <vector>
#include <string>
#include <sstream>

class Solution {
public:
    int findMinDifference(std::vector<std::string>& timePoints) {
        std::vector<int> minutes;
        for (auto tp : timePoints){
            int h, m;
            char temp;
            std::istringstream(tp) >> h >> temp >> m;
            minutes.push_back(60*h + m);
        }

        std::sort(minutes.begin(), minutes.end());
        int ans = 60*24;

        for(int i = 0; i < minutes.size()-1; ++i){
            ans = std::min(ans, minutes[i+1]-minutes[i]);
        }

        ans = std::min(ans, 60*24-minutes.back()+minutes.front());

        return ans;   
    }
};

// Topic: Array, Sorting, String
// Time Complexity: O(n log n)
// Space Complexity: O(n)

File: Python/0539.py
from typing import List

class Solution:
    def findMinDifference(self, timePoints: List[str]) -> int:
        minutes = []

        for tp in timePoints:
            h, m = map(int, tp.split(":"))
            minutes.append(60*h + m)
        
        minutes.sort()
        ans = 60*24
        for i in range(len(minutes)-1):
            ans = min(ans, minutes[i+1]-minutes[i])

        ans = min(ans, 24*60-minutes[-1] + minutes[0])

        return ans


# Topic: Array, Sorting, String
# Time Complexity: O(nlogn)
# Space Complexity: O(n)